### PR TITLE
ncdns: 2020-11-22 -> 2021-07-18

### PR DESCRIPTION
**Pull request by rnhmjoj**
###### Motivation for this change

Mainly switching to `buildGoModule`

###### Things done

- Built on platform(s)
  - [x] x86_64-linux (NixOS)
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [x] Tested via `ncdns.tests`
- [x] Tested compilation of all packages that depend on this change
- [x] Tested execution of all binary files
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 137137` at 77bc057a run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built successfully:</summary>
  <ul>
    <li>ncdns</li>
  </ul>
</details>
<details>
  <summary>1 suggestion:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/patch-phase.md">patch-phase</a>

`patchPhase` should not be overridden, use `postPatch` instead.

Near pkgs/servers/dns/ncdns/default.nix:82:3:
```
   |
82 |   patchPhase = ''
   |   ^
```
</li>
  </ul>
</details>

**Review by c00w — APPROVED** [review]
FYI - I don't have commit privileges so my approval is probably worthless :(

But this is crazy impressive and I'm both impressed and curious as to what ncdns is doing that makes it take this much work to get it to build.

File: pkgs/servers/dns/ncdns/default.nix [hunk 1]
```diff
@@ -1,27 +1,104 @@
...
+      cp -r --no-preserve=mode "${ncdns}" "$out/ncdns"
+    '';
   };
 
-  patches = [ ./fix-tpl-path.patch ];
+in
+
+buildGoModule {
+  pname = "ncdns";
+  version = "2020-07-18";
```
Comment (SuperSandro2000) [inline pkgs/servers/dns/ncdns/default.nix:34]:
```suggestion
  version = "unstable-2020-07-18";
```

File: pkgs/servers/dns/ncdns/default.nix [hunk 2]
```diff
@@ -1,27 +1,104 @@
...
+
+  vendorSha256 = "08g1gdbg9svhf475572l27pkj4aniga3j6ij1wznwjylx98sygfh";
+
+  # Override the go-modules fetcher derivation to apply
+  # upstream's patch of the crypto/x509 library.
+  modBuildPhase = ''
+    go mod init github.com/namecoin/x509-compressed
+    go mod tidy
+    go generate ./...
+    go mod tidy
```
[the remark this answers is not in the document]
Comment (rnhmjoj) [reply pkgs/servers/dns/ncdns/default.nix:46]: I'm following the build instructions very closely, here.
I haven't checked this particular line, but they seemed on point: in fact I tried doing things differently a couple of times and it didn't go well.

File: pkgs/servers/dns/ncdns/default.nix [hunk 3]
```diff
@@ -1,27 +1,104 @@
...
+
+    cd ncdns
+    go mod init github.com/namecoin/ncdns
+    go mod edit \
+      -replace github.com/coreos/go-systemd=github.com/coreos/go-systemd/v22@latest \
+      -replace github.com/namecoin/x509-compressed=/build/source
```
[the remark this answers is not in the document]
Comment (rnhmjoj) [reply pkgs/servers/dns/ncdns/default.nix:52]: Thanks, I was in fact looking for this variable.

File: pkgs/servers/dns/ncdns/default.nix [hunk 4]
```diff
@@ -1,27 +1,104 @@
...
+  '';
+
+  preBuild = ''
+    chmod -R u+w vendor
+    mv -t . vendor/go.{mod,sum}
+  '';
+
+  preCheck = ''
+    # needed to run the ncdns test suite
+    ln -s $PWD/vendor ../../go/src
```
[the remark this answers is not in the document]
Comment (rnhmjoj) [reply pkgs/servers/dns/ncdns/default.nix:95]: Nope, this doesn't work.

File: pkgs/servers/dns/ncdns/default.nix (hunk 2 above)
[the remark this answers is not in the document]
Comment (rnhmjoj) [reply pkgs/servers/dns/ncdns/default.nix]: Apparently works with one `tidy` too.

**Comment by rnhmjoj** [comment]
> FYI - I don't have commit privileges so my approval is probably worthless :(

Not at all, from your commit history I'd say you're much more knowledgeable in Go packaging than I am. Thanks for looking at this!

**Pull request merged**